src/gui/log/logmodel.cpp

```diff
@@ -0,0 +1,179 @@
+ */
+
+#include "logmodel.h"
+
+#include <QApplication>
+#include <QDateTime>
+#include <QColor>
+#include <QPalette>
+
+#include "base/logger.h"
+
+BaseLogModel::BaseLogModel(int initialSize, QObject *parent)
+    : QAbstractListModel(parent)
+    , m_msgCount(initialSize)
+    , m_cache(64)
+{
+    if (m_msgCount == 0)
+        m_startIndex = 0;
+}
+
+int BaseLogModel::rowCount(const QModelIndex &) const
+{
+    if (m_startIndex == -1)
+        return 0;
+    return m_msgCount;
+}
+
+int BaseLogModel::columnCount(const QModelIndex &) const
+{
+    return 1;
+}
+
+QVariant BaseLogModel::data(const QModelIndex &index, const int role) const
+{
+    if (!index.isValid() || (m_msgCount == 0))
+        return {};
+
+    const int i = m_msgCount - index.row() + m_startIndex - 1;
+
+    Item *item = m_cache[i];
+    if (!item) {
+        item = new Item{rowData(i)};
+        m_cache.insert(i, item, 1);
+    }
+
+    switch (role) {
+    case Qt::DisplayRole:
+        return item->displayRole;
+    case Qt::ForegroundRole:
+        return item->foregroundRole;
+    case Qt::UserRole:
+        return item->userRole;
+    default:
+        return {};
+    }
+}
+
+void BaseLogModel::addNewMessage(int index)
+{
+    if (m_startIndex == -1)
+        m_startIndex = index;
+
+    beginInsertRows(QModelIndex(), 0, 0);
+    ++m_msgCount;
+    endInsertRows();
+
+    const int count = rowCount();
+    if (count > MAX_LOG_MESSAGES) {
+        const int lastMessage = count - 1;
+        beginRemoveRows(QModelIndex(), lastMessage, lastMessage);
+        --m_msgCount;
+        endRemoveRows();
+    }
+}
+
+void BaseLogModel::reset()
+{
+    beginResetModel();
+    m_startIndex = m_msgCount;
+    m_msgCount = 0;
+    endResetModel();
+}
+
+LogMessageModel::LogMessageModel(QObject *parent) : 
+    BaseLogModel(Logger::instance()->messageCount(), parent)
+{
+    connect(Logger::instance(), &Logger::newLogMessage, this, &LogMessageModel::handleNewMessage);
+}
+
+void LogMessageModel::handleNewMessage(const Log::Msg &msg)
+{
+    addNewMessage(msg.id);
+}
+
+BaseLogModel::Item LogMessageModel::rowData(const int index) const
+{
+    const Log::Msg msg = Logger::instance()->message(index);
+
+    Item item;
+    const QDateTime time = QDateTime::fromMSecsSinceEpoch(msg.timestamp);
+    item.displayRole = QString::fromLatin1("%1 - %2").arg(time.toString(Qt::SystemLocaleShortDate), msg.message);
+
+    switch (msg.type) {
+    // The RGB QColor constructor is used for performance
+    case Log::NORMAL:
+        item.foregroundRole = QApplication::palette().color(QPalette::WindowText);
+        break;
+
+    case Log::INFO:
+        item.foregroundRole = QColor(0, 0, 255); // blue
+        break;
+
+    case Log::WARNING:
+        item.foregroundRole = QColor(255, 165, 0);  // orange
+        break;
+
+    case Log::CRITICAL:
+        item.foregroundRole = QColor(255, 0, 0);  // red
+        break;
+
+    default:
+        Q_ASSERT(false && "invalid message type");
```

Comment:
What do you mean by using `&&` here?? or you should use `Q_ASSERT_X`

Reply:
removed the string